repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -0,0 +1,138 @@
+
+        if (parentDomainObject == null) {
+            entity.removeAttribute(PARENT_DOMAIN_QN);
+            entity.removeAttribute(SUPER_DOMAIN_QN);
+        } else {
+            AtlasVertex parentDomain = retrieverNoRelation.getEntityVertex(parentDomainObject);
+            String parentDomainQualifiedName = parentDomain.getProperty(QUALIFIED_NAME, String.class);
+
+            entity.setAttribute(PARENT_DOMAIN_QN, parentDomainQualifiedName);
+
+            String[] splitted = parentDomainQualifiedName.split("/");
+            String superDomainQualifiedName = String.format("%s/%s/%s", splitted[0], splitted[1], splitted[2]);
+            entity.setAttribute(SUPER_DOMAIN_QN, superDomainQualifiedName);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void processUpdateProduct(AtlasEntity entity, AtlasVertex vertex, EntityMutationContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateProduct");
+
+        AtlasEntity storedProduct = entityRetriever.toAtlasEntity(vertex);
+        AtlasObjectId currentParent = (AtlasObjectId) storedProduct.getRelationshipAttribute(PRODUCT_DOMAIN_REL_TYPE);
+
+        AtlasEntityHeader newParent = getNewParentDomain(entity, context);
+
+        if (currentParent == null && newParent != null) {
+            // attaching parent for the first time, set attributes as well
+
+            String parentDomainQualifiedName = (String) newParent.getAttribute(QUALIFIED_NAME);
+            entity.setAttribute(PARENT_DOMAIN_QN, parentDomainQualifiedName);
+
+            String superDomainQualifiedName = (String) newParent.getAttribute(SUPER_DOMAIN_QN);
+            if (StringUtils.isEmpty(superDomainQualifiedName)) {
+                String[] splitted = parentDomainQualifiedName.split("/");
+                superDomainQualifiedName = String.format("%s/%s/%s", splitted[0], splitted[1], splitted[2]);
+            }
+
+            entity.setAttribute(SUPER_DOMAIN_QN, superDomainQualifiedName);
+
+        } else {
+            //never update these attributes with API request
+            entity.removeAttribute(PARENT_DOMAIN_QN);
+            entity.removeAttribute(SUPER_DOMAIN_QN);
+            if (entity.getRelationshipAttributes() != null) {
+                entity.getRelationshipAttributes().remove(PRODUCT_DOMAIN_REL_TYPE);
+            }
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private AtlasEntityHeader getNewParentDomain(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("ProductPreProcessor.getNewParentDomain");
+
+        AtlasObjectId objectId = (AtlasObjectId) entity.getRelationshipAttribute(PRODUCT_DOMAIN_REL_TYPE);
+
```

Comment:
The object returned by entity.getRelationshipAttribute(PRODUCT_DOMAIN_REL_TYPE) is instanceOf Map and not AtlasObjectId in my opinion. Can you please check it once.